### ISSUE TITLE
Web: 3.19 tracking issue

### ISSUE THREAD

**Issue by lguychard** [sourcegraph/sourcegraph#12293]
## Iteration name

"We want to experience the pain"

### Use recent improvements from design to dogfood and improve our design-to-developer handoff process

**Problem**

As we are building out our design team, our most recent data points on design-to-developer handoff have outlined pain points. We want to use recent improvements from design that are ready for handoff ([headers & breadcrumbs for common page types](https://github.com/sourcegraph/sourcegraph/issues/11774), [information architecture improvements](https://github.com/sourcegraph/sourcegraph/issues/11775)) to dogfood our interface between design and development, and improve it in any way we can.

**Planned work**
- The team will collectively review the designs for #11774, #11775
- We'll derive an implementation plan, making sure that we all get exposure to the handoff process.
- We'll take this opportunity to identify what is missing / what needs to be cleand up in Storybook (which is meant to be the developer-facing interface to our component system), and take action there
- During implementation, we'll collect data points about the handoff:
    - We'll identify the pain points that we can fix. We'll fix them immediately if possible, otherwise, we'll formulate the resulting problem statements we want to fix next.
    - We'll also identify what did work, and turn it into documentation for future developers to follow.

### Reduce our reliance on the e2e regression test suite to the bare minimum

**Problem**

Our heavy reliance on our end-to-end regression test suite creates pain at the end of every release cycle: regression tests are run manually, which is a painful process, and they require incremental fixes at the end of every release cycle (https://github.com/sourcegraph/sourcegraph/pull/12258, https://github.com/sourcegraph/sourcegraph/pull/12439). Yet, our [testing pyramid](https://about.sourcegraph.com/handbook/engineering/testing#testing-pyramid) documents that we should use end-to-end tests sparingly. While the web team did build an integration tests platform in 3.18, giving us the ability to continuously test complex UI flows in CI with no flakiness, there is still some work left to do to take this over the finish line.

We want to expand our integration test suite, and delete corresponding e2e regression tests, so that our reliance on e2e regression tests is dramatically reduced by the end of the milestone, bringing us significantly closer to a fully automated release process.

**Planned work**

- @twop, who has owned previous audit efforts, owns identifying all flows that should have integration tests, and distributing them to web team members for implementation
- @twop and @felixfbecker will work on using non-hacky graphQL type generation for the query/response types used in integration tests (this is notably being worked on in #12284)
- @lguychard and @marekweb will work on figuring out how to make integration tests work to test extensions and the browser extension, which is blocked by https://github.com/sourcegraph/sourcegraph/issues/12194. @lguychard's work on writing our netflix/pollyjs adapter is relevant here. This will unlock:
    - The ability to test UI flows contributed by extensions (notably our code intelligence UI flows)
    - The ability to test native integrations (which are currently untested when releasing Sourcegraph, as they're not covered by the e2e regression test suite)

## Availability

Period is from July 20th to August 19th (23 working days).

@lguychard: 20d (off starting August 19th)
@felixfbecker: 23d
@marekweb: 23d
@twop: 23d

## Issues in the milestone

The primary source of truth for what the team is working on in this milestone is the **Goals** section above. Progress on those goals is communicated through weekly updates as comments to this tracking issue. As such, this section should not be taken as a comprehensive overview of what the team is working on in 3.19.

@Unassigned: __11.00d__

- [ ] Placeholder (maybe split later): migrate the rest of the areas to the new model according to RFC 155. [#10750](https://github.com/sourcegraph/sourcegraph/issues/10750) __8d__ 
- [ ] Migrate ExtExtensions API to a new model described in RFC 155 [#10749](https://github.com/sourcegraph/sourcegraph/issues/10749) __3d__ 

@felixfbecker

- [x] Adopt graphql-code-generator [#12559](https://github.com/sourcegraph/sourcegraph/pull/12559) :shipit:

@lguychard

- [ ] Add Repository / Update Repository actions unclear [#10899](https://github.com/sourcegraph/sourcegraph/issues/10899) 🐛

@limitedmage

- [x] Add tags to Git-extras extension [#12485](https://github.com/sourcegraph/sourcegraph/issues/12485) 

@marekweb

- [x] Bitbucket Server native integration prevents file header text from wrapping [#11110](https://github.com/sourcegraph/sourcegraph/issues/11110) 🐛

@tjkandala

- [x] Browser ext links to Sourcegraph don't appear on repos when branches specified [#11057](https://github.com/sourcegraph/sourcegraph/issues/11057) 🐛

**Comment by felixfbecker** [comment]
**Last week** I worked on fixing issues with integration tests that poppoed up and making some more improvements. Besides that, I met with members of the design team and worked on tooling to solve problems I identified in our design-engineering workflow, such as adding Chromatic to give designers the ability to view our components (Storybook) and review visual changes and integrating Figma into our Storybook.
**This week** I want to plan this iteration and perhaps work on remaining RFC155 work.

**Comment by lguychard** [comment]
**Last week** I worked with the team to define goals for the iteration (and on our goal-setting process in general), and prepared the [onboarding plan](https://github.com/sourcegraph/about/pull/1259) for our new teammate joining Monday.

**Comment by marekweb** [comment]
**Last week** I worked on wrapping up loose ends of my integration tests. I made contributions to the developer documentation and to the handbook. I collected notes for more additions that I'd like to make to the docs, I'm looking forward to making more contributions there.

I followed up on the review of our Sourcegraph for Firefox add-on and I'm hoping to accelerate the process, because it feels like it's getting very delayed.

We spent time this week as a team to define goals for the iteration.

**Next week** I plan to work on the problem of how we can run integration puppeteer tests on extensions and browser extensions. I also want to get familiar with all the components we have in Storybook and find out about the issues there.

**Comment by twop** [comment]
**Last week** I worked on migrating to https://graphql-code-generator.com/. I was able to successfully generate operation types with it that we use for tests, but I haven't integrated it into our build pipeline which has `watch` and `build` modes. Generating operation types alone takes 5 seconds and significant CPU load (enough for my fans to go wild), and that makes me hesitant to use it in watch mode. I was able to make a draft implementation of watching only for files that have `gql`  tags in them but I still think it is not good enough DevEx.

As a part of this effort I tried to migrate from gql2ts library and use graphql-code-generator to generate schema.ts (schema ts types) too but after working on it for a day I realized it probably makes more sense to do this work in a separate PR, and for graphql-operations.ts files just inline schema types as a temporary solution (otherwise there are type errors).

**Next week** I'm looking forward to our planned pairing(tetragramming?) session to kick off 3.19 work. Also I need to wrap up the [PR](https://github.com/sourcegraph/sourcegraph/pull/12284) and hopefully finish the work described above to fully migrate to graphql-code-generator

**Comment by felixfbecker** [comment]
**Last week** I worked on [Alicja's badge redesign](https://github.com/sourcegraph/sourcegraph/pull/12390), updates to our info color and related Storybook updates. I spent time with the team to define our 3.19 goals and thought about how we can make sure we all have the context for that, e.g. by making [recurring issue triages](https://github.com/sourcegraph/about/pull/1260) part of our process. I also finally added  documentation to the handbook on [how we style UI at Sourcegraph](https://github.com/sourcegraph/about/pull/1237).
**This week** we'll kick off 3.19 work in quadgramming sessions.

**Comment by lguychard** [comment]
**Last week** I worked with the team on the breadcrumbs project: reviewing the designs as a group, then implementing them, starting with using a new shared breadcrumbs component instead of the ad-hoc breadcrumbs component that had been used in the repository header. We've been making progress on this PR: https://github.com/sourcegraph/sourcegraph/pull/12536

**Comment by felixfbecker** [comment]
**Last week** I worked to bring the new GraphQL codegen over the finish line and worked on the breadcrumbs project.

**Comment by marekweb** [comment]
**Last week* I worked on the breadcrumbs project, made some contributions to page headers to try to make them into a reusable component for all the pages.

I did a dive into the design work in Figma for web styles and components, to plan how we can implement them as actual components and I added some content into Storybook, namely for the tentative PageHeader.

**This week** I'd like to shift my work back towards integration tests, because this fits into the work needed for the 3.19 release which I'll do as release captain. I want to get some work done on this early ahead of the release so that there's time to smooth out any issues along the way.

**Comment by tjkandala** [comment]
**Last week** I joined Sourcegraph and set up my development environment. I worked on a [bug fix for the browser extension](https://github.com/sourcegraph/sourcegraph/pull/12511) and [fault tolerance for loading extensions](https://github.com/sourcegraph/sourcegraph/pull/12573). I also found and fixed a long-standing performance issue in the extension registry and proposed some long-term improvements to the web app architecture (#12637).

**This week** I'll be working to [improve the discoverability of Sourcegraph extensions](https://github.com/sourcegraph/sourcegraph/issues/11872). I will also work on [integration tests for add/update/remove repository actions](https://github.com/sourcegraph/sourcegraph/issues/12522) and the extension registry (since I have made and will continue to make changes to that part of the codebase). I want to allocate some time to work on long-term solutions to the performance issues discussed in #12637

**Comment by limitedmage** [comment]
*Last week:* I set up all my accounts and dev environment. I had pairing sessions with everyone on the team to get familiar with what people are working on. I fixed bug #12721 where interactive search doesn't play nice with proposed queries coming from the server and learned to build unit tests and integration tests as part of this work. I [created and published an extension](https://sourcegraph.com/extensions/sourcegraph/open-in-editor) to address #10093 so users can open the viewed file in the editor of their choice. 

*This week:* I will continue learning more about the project. I plan to tackle issue #8498 so that the browser extension respects the user's default pattern type setting. Once this is done, I'll look to see what I issues I can assist working on based on the iteration's priorities.

**Comment by tjkandala** [comment]
**Last week** I wrote integration tests for repository actions and the extension registry. I finished refactoring the extension registry and implemented the filtering functionality specified in RFC 209. (Draft PR: https://github.com/sourcegraph/sourcegraph/pull/12810/files)

**This week** I will continue working on RFC 209 (mostly style changes remain), finish refactoring `SourcegraphWebApp` for performance and maintainability improvements, and learn about the extension API architecture detailed in RFC 155

**Comment by lguychard** [comment]
**Last week**  I pitched in on the [breadcrumbs & headers PR](https://github.com/sourcegraph/sourcegraph/pull/12536), which we should land this week. I also investigated using [puppeteer-interceptor-poc](https://github.com/sourcegraph/puppeteer-interceptor-poc) to investigate whether we could use it to [unblock testing Sourcegraph extensions](https://github.com/sourcegraph/sourcegraph/issues/12194) -- good news is, we can!

**This week** I'll be wrapping up any loose ends before going on holiday August 15-30.

**Comment by marekweb** [comment]
**Last week** I split my time between doing some preparation for captaining the 3.19 release, getting familiar with release scripts and pairing with @uwedeportivo to get access to his extensive knowledge. I did some work on integration tests and followed up on @lguychard's work on puppeteer interception, although I didn't get to the point of building on top of the proof-of-concept yet.

**This week** I plan to allocate my time between contributing to finishing the breadcrumbs/headers project, possibly working on puppeteer interception and trying out testing Sourcegraph extensions, starting the release tasks for 3.19, and also moving things forward on getting our Firefox add-on re-published since we now have some new specific action points to address to comply with add-on policy.

**Comment by marekweb** [comment]
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.19 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you

**Comment by marekweb** [comment]
**Last week** I made the branch cut for 3.19 as release captain, I worked on the Firefox add-on update, and I spent some time on integration tests although I didn't make much progress there.

**This week** I'm aiming to finish the Firefox add-on, I'm working towards the release on Thursday, and I'm doing goal setting with the team. I'm also going to see if I can revisit the testing work.

**Comment by felixfbecker** [comment]
**Last week** I defined our long and medium term goals with the team and helped to wrap up the breadcrumbs/PageHeader PR #12536.

**This week** I will review the RFC ([RFC 209](https://docs.google.com/document/d/19f4xleYBU1zZZdqMmXlLmFxeR-fwEpOwTOgViOFOnyo/edit#heading=h.trqab8y0kufp)) and designs for extension discoverability, as well as kick off the browser extension discoverability project with the team and design ([RFC 211](https://docs.google.com/document/d/19f4xleYBU1zZZdqMmXlLmFxeR-fwEpOwTOgViOFOnyo/edit#heading=h.trqab8y0kufp)). I'll also field the work for the sign-in/sign-out project. I'll make sure we have GitHub issues and assignees for all sub-tasks this iteration (this came up in our retro).
